dvc/external_repo.py

```diff
@@ -65,66 +68,211 @@ def clean_repos():
+                    else:
+                        raise RecursiveImportError(
+                            path_info.relative_to(self.root_dir)
+                        )
+        return fetch_infos
+
+    def _fetch_info(self, path_info):
+        if self.repo_tree.isdvc(path_info):
+            (out,) = self.find_outs_by_path(path_info, strict=False)
+            filter_info = path_info.relative_to(out.path_info)
+        else:
+            filter_info = None
+        return path_info, filter_info
+
+    def get_external(self, path, to_info, recursive=False, **kwargs):
         """
         Pull the corresponding file or directory specified by `path` and
-        checkout it into `to_info`.
+        into `to_info`.
 
         It works with files tracked by Git and DVC, and also local files
         outside the repository.
         """
-        out = None
         path_info = PathInfo(self.root_dir) / path
 
-        with suppress(OutputNotFoundError):
-            (out,) = self.find_outs_by_path(fspath(path_info), strict=False)
-
-        try:
-            if out and out.use_cache:
-                self._pull_cached(out, path_info, to_info)
-                return
+        if not self.repo_tree.exists(path_info):
+            raise PathMissingError(path, self.url)
 
-            # Check if it is handled by Git (it can't have an absolute path)
-            if os.path.isabs(path):
-                raise FileNotFoundError
+        fetch_infos = [self._fetch_info(path_info)]
+        if self.repo_tree.isdir(path_info) and not self.repo_tree.isdvc(
+            path_info
+        ):
+            fetch_infos.extend(self._recursive_outputs(path_info, recursive))
 
-            fs_copy(fspath(path_info), fspath(to_info))
-        except FileNotFoundError:
-            raise PathMissingError(path, self.url)
+        self._fetch_external(fetch_infos, **kwargs)
 
-    def _pull_cached(self, out, path_info, dest):
         with self.state:
-            tmp = PathInfo(tmp_fname(dest))
-            src = tmp / path_info.relative_to(out.path_info)
+            self.repo_tree.copytree(path_info, to_info)
 
-            out.path_info = tmp
+    def fetch_external(self, files, **kwargs):
+        """Fetch erepo files into the specified cache.
 
-            # Only pull unless all needed cache is present
-            if out.changed_cache(filter_info=src):
-                self.cloud.pull(out.get_used_cache(filter_info=src))
+        Works with files tracked by Git and DVC.
+        """
+        files = [(PathInfo(self.root_dir) / name, None) for name in files]
+        return self._fetch_external(files, **kwargs)
+
+    def _fetch_external(self, fetch_infos, **kwargs):
+        downloaded, failed = 0, 0
 
+        with self.state:
+            for path_info, filter_info in fetch_infos:
+                if self.repo_tree.isdvc(path_info):
+                    (out,) = self.find_outs_by_path(path_info, strict=False)
```

Comment:
Was kinda hoping we no longer need `find_outs_by_path` and could use tree instead.

Reply:
so something like `RepoTree.outs_for_path(...)`? and if the path isn't an out just raise OutputNotFoundError?

Reply:
@pmrowla No, I actually thought about something native that would possibly abstract it away so we don't have to care about it being an output in this code at all, having tree handle all of that for us. Something like `self.cache.local.save_obj(tree.open(path))`. Obviously we'll need to somehow handle fetching inside of the tree (or rather somehow ask it what cache it needs us to download for specific path), so I'm not completely sure if this is doable in a nice way.